### ISSUE TITLE
HandlerManager throws ArrayIndexOutOfBoundsException under high load

### ISSUE THREAD

**Issue by PennyYip** [eclipse-vertx/vert.x#2783]
Upon stress on websocket connectivity (Connect and disconnect quickly), HandlerManager is throwing ArrayIndexOutOfBoundsException as below - 

```
[io.vertx.core.impl.ContextImpl]  Unhandled exception
java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 1
        at java.base/java.util.concurrent.CopyOnWriteArrayList.elementAt(CopyOnWriteArrayList.java:386)
        at java.base/java.util.concurrent.CopyOnWriteArrayList.get(CopyOnWriteArrayList.java:399)
        at io.vertx.core.net.impl.HandlerManager$Handlers.chooseHandler(HandlerManager.java:85)
        at io.vertx.core.net.impl.HandlerManager.chooseHandler(HandlerManager.java:50)
        at io.vertx.core.http.impl.HttpServerImpl$ServerHandlerWithWebSockets.handshake(HttpServerImpl.java:724)
        at io.vertx.core.http.impl.HttpServerImpl$ServerHandlerWithWebSockets.handleMessage(HttpServerImpl.java:701)
        at io.vertx.core.http.impl.HttpServerImpl$ServerHandlerWithWebSockets.handleMessage(HttpServerImpl.java:619)
        at io.vertx.core.net.impl.VertxHandler.lambda$channelRead$1(VertxHandler.java:146)
        at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:337)
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

vertx-core version: 3.5.3

**Comment by vietj** [comment]
thanks for reporting this

**Comment by vietj** [comment]
it seems that your server is executed by a worker verticle, right ?

**Comment by vietj** [comment]
can you try with latest 3.6.2 ? I believe the issue should not happen with this version.

**Comment by PennyYip** [comment]
Thanks @vietj for getting back quickly, I shall upgrade and test. Will let you know the result shortly.

**Comment by vietj** [comment]
thanks @PennyYip this part has been improved in 3.6 and the HandlerManager is not called anymore on a websocket upgrade which is why it should not happen anymore

**Comment by PennyYip** [comment]
@vietj 
I have upgrade vertx-core to 3.6.2, the ArrayIndexOutOfBoundsException is gone. However, a NPE appears occasionally.

I have not change any code I wrote, only upgraded vertx. 

```
Unhandled exception
java.lang.NullPointerException
        at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:44)
        at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:29)
        at io.vertx.core.http.impl.HttpServerRequestImpl.handleBegin(HttpServerRequestImpl.java:145)
        at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:130)
        at io.vertx.core.net.impl.ConnectionBase.handleRead(ConnectionBase.java:386)
        at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
        at io.vertx.core.impl.WorkerContext.lambda$wrapTask$0(WorkerContext.java:34)
        at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by vietj** [comment]
ok that's good to know and likely another that you found.

I will try to reproduce this one.

Julien

**Comment by vietj** [comment]
would you mind trying to create a reproducer for it ?

also more details : how often is occasionally ? how many concurrent ws do you process ? etc...

**Comment by vietj** [comment]
also, who is disconnecting ? the client ?

**Comment by vietj** [comment]
reproducer would be helpful, I've tried to make one unsucesfully

**Comment by vietj** [comment]
so the best workaround is that you set an request handler on the HttpServer and send a 400 response, you might also want to investigate what those requests are.

**Comment by vietj** [comment]
For the record the issue has been fixed https://github.com/eclipse-vertx/vert.x/issues/2785

**Comment by vietj** [comment]
Currently in 3.5.x , won't fix as now we support 3.6
